# Laboratory excercise 5

The notebook contains excercise connected to the auditory excerice 6. For any questions feel free to contact assistant: eda.jovicic@fer.hr

In [1]:
import pandas as pd
import numpy as np
import statsmodels.formula.api as smf

# set seed for consistency
np.random.seed(2)

1. Load the cleaned dataset from the first exercise. If you haven't saved the dataset, rerun the exercise and save the final dataset.

In [2]:
#load data

df = pd.read_csv('students_score_handled.csv', index_col = 0)

In [3]:
df

EthnicGroup  ParentEduc  TestPrep  ParentMaritalStatus  PracticeSport  \
0                2           1         1                    1              1   
1                2           5         1                    1              2   
2                1           3         1                    2              2   
3                0           0         1                    1              0   
4                2           5         1                    1              2   
...            ...         ...       ...                  ...            ...   
29112            3           2         1                    2              2   
29113            4           2         1                    2              1   
29114            2           2         0                    1              2   
29115            3           0         0                    1              1   
29116            1           5         1                    1              0   

       IsFirstChild  NrSiblings  TransportMeans  WklyStudyHours  MathScore  \
0                 1         3.0               1               1         71   
1                 1         0.0               2               0         69   
2                 1         4.0               1               1         87   
3                 0         1.0               2               0         45   
4                 1         0.0               1               0         76   
...             ...         ...             ...             ...        ...   
29112             0         2.0               1               0         59   
29113             0         1.0               0               0         58   
29114             0         1.0               0               0         61   
29115             0         3.0               1               0         82   
29116             0         1.0               1               0         64   

       ReadingScore  WritingScore  Gender_female  Gender_male  \
0                71            74            1.0          0.0   
1                90            88            1.0          0.0   
2                93            91            1.0          0.0   
3                56            42            0.0          1.0   
4                78            75            0.0          1.0   
...             ...           ...            ...          ...   
29112            61            65            1.0          0.0   
29113            53            51            0.0          1.0   
29114            70            67            1.0          0.0   
29115            90            93            1.0          0.0   
29116            60            58            0.0          1.0   

       LunchType_free/reduced  LunchType_standard  
0                         0.0                 1.0  
1                         0.0                 1.0  
2                         0.0                 1.0  
3                         1.0                 0.0  
4                         0.0                 1.0  
...                       ...                 ...  
29112                     0.0                 1.0  
29113                     0.0                 1.0  
29114                     1.0                 0.0  
29115                     0.0                 1.0  
29116                     0.0                 1.0  

[29117 rows x 16 columns]

2. Using linear regression, model the MathScore using features: TestPrep, NrSiblings, ParentMaritalStatus and ReadingScore. 

In [4]:
# Declares the model
# TestPrep - categorical variable (0 or 1)
#  looking at the original dataset - Label Encoder assigns labels following alphabetical order of data (none - 1, completed - 0)
# NrSiblings - discrete 
# ParentMaritalStatus - categorical (0 (divorced), 1 (married), 2 (single), 3 (widowed))
# ReadingScore - discrete


mod = smf.ols(formula='MathScore ~ C(TestPrep) + NrSiblings + C(ParentMaritalStatus) + ReadingScore', data=df)

res = mod.fit()

res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              MathScore   R-squared:                       0.670
Model:                            OLS   Adj. R-squared:                  0.669
Method:                 Least Squares   F-statistic:                     9829.
Date:                Sat, 06 Jan 2024   Prob (F-statistic):               0.00
Time:                        01:20:12   Log-Likelihood:            -1.0471e+05
No. Observations:               29117   AIC:                         2.094e+05
Df Residuals:                   29110   BIC:                         2.095e+05
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
===============================================================================================
                                  coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
Intercept                       6.2302      0.314     19.867      0.000       5.616       6.845
C(TestPrep)[T.1]                1.1060      0.113      9.807      0.000       0.885       1.327
C(ParentMaritalStatus)[T.1]     0.1733      0.145      1.193      0.233      -0.112       0.458
C(ParentMaritalStatus)[T.2]    -0.1318      0.168     -0.786      0.432      -0.460       0.197
C(ParentMaritalStatus)[T.3]     0.5432      0.396      1.373      0.170      -0.232       1.319
NrSiblings                      0.0155      0.036      0.428      0.668      -0.055       0.086
ReadingScore                    0.8570      0.004    239.510      0.000       0.850       0.864
==============================================================================
Omnibus:                      233.486   Durbin-Watson:                   1.987
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              155.618
Skew:                          -0.010   Prob(JB):                     1.61e-34
Kurtosis:                       2.642   Cond. No.                         562.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Questions:

a) Using the summary, write down the formula with coefficients

$MathScore = 6.2302 + 1.1060 * TestPrep + 0.1733 * ParentMaritalStatus[T.1] + (-0.1318) * ParentMaritalStatus[T.2] + 0.5432 * ParentMaritalStatus[T.3] + 0.0155 * NrSiblings + 0.8570 * ReadingScore$


In [5]:
# I reversed the labels of TestPrep so that "completed" is 1, and "none" is 0 and tried the model like that
df['TestPrep'] = 1 - df['TestPrep']

In [6]:
df

EthnicGroup  ParentEduc  TestPrep  ParentMaritalStatus  PracticeSport  \
0                2           1         0                    1              1   
1                2           5         0                    1              2   
2                1           3         0                    2              2   
3                0           0         0                    1              0   
4                2           5         0                    1              2   
...            ...         ...       ...                  ...            ...   
29112            3           2         0                    2              2   
29113            4           2         0                    2              1   
29114            2           2         1                    1              2   
29115            3           0         1                    1              1   
29116            1           5         0                    1              0   

       IsFirstChild  NrSiblings  TransportMeans  WklyStudyHours  MathScore  \
0                 1         3.0               1               1         71   
1                 1         0.0               2               0         69   
2                 1         4.0               1               1         87   
3                 0         1.0               2               0         45   
4                 1         0.0               1               0         76   
...             ...         ...             ...             ...        ...   
29112             0         2.0               1               0         59   
29113             0         1.0               0               0         58   
29114             0         1.0               0               0         61   
29115             0         3.0               1               0         82   
29116             0         1.0               1               0         64   

       ReadingScore  WritingScore  Gender_female  Gender_male  \
0                71            74            1.0          0.0   
1                90            88            1.0          0.0   
2                93            91            1.0          0.0   
3                56            42            0.0          1.0   
4                78            75            0.0          1.0   
...             ...           ...            ...          ...   
29112            61            65            1.0          0.0   
29113            53            51            0.0          1.0   
29114            70            67            1.0          0.0   
29115            90            93            1.0          0.0   
29116            60            58            0.0          1.0   

       LunchType_free/reduced  LunchType_standard  
0                         0.0                 1.0  
1                         0.0                 1.0  
2                         0.0                 1.0  
3                         1.0                 0.0  
4                         0.0                 1.0  
...                       ...                 ...  
29112                     0.0                 1.0  
29113                     0.0                 1.0  
29114                     1.0                 0.0  
29115                     0.0                 1.0  
29116                     0.0                 1.0  

[29117 rows x 16 columns]

In [7]:
mod = smf.ols(formula='MathScore ~ C(TestPrep) + NrSiblings + C(ParentMaritalStatus) + ReadingScore', data=df)

res = mod.fit()

res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              MathScore   R-squared:                       0.670
Model:                            OLS   Adj. R-squared:                  0.669
Method:                 Least Squares   F-statistic:                     9829.
Date:                Sat, 06 Jan 2024   Prob (F-statistic):               0.00
Time:                        01:21:50   Log-Likelihood:            -1.0471e+05
No. Observations:               29117   AIC:                         2.094e+05
Df Residuals:                   29110   BIC:                         2.095e+05
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
===============================================================================================
                                  coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
Intercept                       7.3362      0.287     25.521      0.000       6.773       7.900
C(TestPrep)[T.1]               -1.1060      0.113     -9.807      0.000      -1.327      -0.885
C(ParentMaritalStatus)[T.1]     0.1733      0.145      1.193      0.233      -0.112       0.458
C(ParentMaritalStatus)[T.2]    -0.1318      0.168     -0.786      0.432      -0.460       0.197
C(ParentMaritalStatus)[T.3]     0.5432      0.396      1.373      0.170      -0.232       1.319
NrSiblings                      0.0155      0.036      0.428      0.668      -0.055       0.086
ReadingScore                    0.8570      0.004    239.510      0.000       0.850       0.864
==============================================================================
Omnibus:                      233.486   Durbin-Watson:                   1.987
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              155.618
Skew:                          -0.010   Prob(JB):                     1.61e-34
Kurtosis:                       2.642   Cond. No.                         560.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The formula with coefficients now is

$MathScore = 7.3362 + (-1.1060) * TestPrep + 0.1733 * ParentMaritalStatus[T.1] + (-0.1318) * ParentMaritalStatus[T.2] + 0.5432 * ParentMaritalStatus[T.3] + 0.0155 * NrSiblings + 0.8570 * ReadingScore$


b) Are there any statisticaly important features?

When p value is less than or equal to 0.05, feature is statistically significant. Then, there is a 5 % (or less) chance that the feature (independent variable) has no affect on the dependent variable, and that the results are produced by chance.

TestPrep and ReadingScore have a p value < 0.001 so they can be interpreted as statistically significant.


c) Let's interpret the data, if students do test preparation, how does that affect the outcome on the math exam? 

If students do test preparation, they will have smaller MathScore by 1.1060 than those who do not prepare. With originally altered dataset, Label Encoder assigned 1 to "none" (no preparation) and 0 to "completed" (with preparation) because Label Encoder assigns labels following alphabetical order of data. The same conclusion can be drawn when I reversed the labels of TestPrep so that "completed" means 1, and "none" means 0. Maybe preparation was needed for students with lower scores and lower interest in general so they were the ones with the lower score on the exam in the end.

$MathScore = 6.2302 + 1.1060 * TestPrep$

$MathScore = 7.3362 + (-1.1060) * TestPrep$ (with reversed labels)

d) Are students who practice sport better at math exams or worse? Using linear regression, model the MathScore using only one feature: PracticeSport. Don't include intercept. Using summary, decide which students do the best at math exams - those who never practice sport, sometimes practice sport or those who do it regularly?

In [8]:
# Declare the model

# PracticeSport - 3 possible levels: never (0), regularly (1), sometimes (2)
# a categorical variable (also called qualitative variable) is a variable that can take on one of a limited, 
# and usually fixed, number of possible values, assigning each individual or other unit of observation 
# to a particular group or nominal category on the basis of some qualitative property.

# without intercept (-1)

mod2 = smf.ols(formula='MathScore ~ C(PracticeSport)-1', data=df)

res2 = mod2.fit()

res2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              MathScore   R-squared:                       0.006
Model:                            OLS   Adj. R-squared:                  0.006
Method:                 Least Squares   F-statistic:                     87.35
Date:                Sat, 06 Jan 2024   Prob (F-statistic):           1.51e-38
Time:                        01:28:51   Log-Likelihood:            -1.2074e+05
No. Observations:               29117   AIC:                         2.415e+05
Df Residuals:                   29114   BIC:                         2.415e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=======================================================================================
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
C(PracticeSport)[0]    64.1576      0.244    262.426      0.000      63.678      64.637
C(PracticeSport)[1]    67.8441      0.150    452.998      0.000      67.551      68.138
C(PracticeSport)[2]    66.2896      0.126    526.520      0.000      66.043      66.536
==============================================================================
Omnibus:                      234.462   Durbin-Watson:                   1.997
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              214.722
Skew:                          -0.171   Prob(JB):                     2.36e-47
Kurtosis:                       2.756   Cond. No.                         1.94
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Best at math exams are students who practice sport regularly (C(PracticeSport)[1]) because the coefficient with this category is the highest (67.8441).